contracts/rust/src/bin/diff_test.rs

```diff
@@ -0,0 +1,1072 @@
+        }
+        Action::PlonkBatchVerify => {
+            let arg1 = cli.arg1.as_ref().expect("Should provide arg1=numProof");
+            let num_proof = arg1.parse::<u32>().unwrap();
+            let (proofs, vks, public_inputs, extra_msgs, _): (
+                Vec<Proof<Bn254>>,
+                Vec<VerifyingKey<Bn254>>,
+                Vec<Vec<Fr>>,
+                Vec<Option<Vec<u8>>>,
+                Vec<usize>,
+            ) = multiunzip(gen_plonk_proof_for_test(num_proof as usize).unwrap());
+
+            // ensure they are correct params
+            let proofs_refs: Vec<&Proof<Bn254>> = proofs.iter().collect();
+            let vks_refs: Vec<&VerifyingKey<Bn254>> = vks.iter().collect();
+            let pi_refs: Vec<&[Fr]> = public_inputs
+                .iter()
+                .map(|pub_input| &pub_input[..])
+                .collect();
+            assert!(PlonkKzgSnark::batch_verify::<SolidityTranscript>(
+                &vks_refs,
+                &pi_refs,
+                &proofs_refs,
+                &extra_msgs
+            )
+            .is_ok());
+
+            let vks_parsed: Vec<ParsedVerifyingKey> = vks.into_iter().map(Into::into).collect();
+            let pis_parsed: Vec<Vec<U256>> = public_inputs
+                .into_iter()
+                .map(|pi| pi.into_iter().map(field_to_u256).collect())
+                .collect();
+            let proofs_parsed: Vec<ParsedPlonkProof> = proofs.into_iter().map(Into::into).collect();
+            let msgs_parsed: Vec<Bytes> = extra_msgs
+                .into_iter()
+                .map(|msg| msg.unwrap().into())
+                .collect();
+
+            let res = (vks_parsed, pis_parsed, proofs_parsed, msgs_parsed);
+            println!("{}", res.encode_hex());
+        }
+        Action::DummyProof => {
+            let mut rng = jf_utils::test_rng();
+            if let Some(arg1) = cli.arg1.as_ref() {
+                let seed = arg1.parse::<u64>().unwrap();
+                rng = StdRng::seed_from_u64(seed);
+            }
+
+            let proof = ParsedPlonkProof::dummy(&mut rng);
+            println!("{}", (proof,).encode_hex());
+        }
+        Action::TestOnly => {
+            let arg1 = cli.arg1.as_ref().expect("Should provide arg1=proof");
+            let proof_parsed = arg1.parse::<ParsedPlonkProof>().unwrap();
+
+            let proof: Proof<Bn254> = proof_parsed.into();
+
+            let res: ParsedPlonkProof = proof.into();
+            println!("{}", (res,).encode_hex());
+        }
+    };
+}
+
+// ------- Helper functions and structs --------
+// ---------------------------------------------
+
+// constant in hex string copied from hardcoded constants from solidity contracts
+
+// TODO: (alex) change to simply using `MontFp!("0x..")` after
+// <https://github.com/arkworks-rs/algebra/pull/635> is on a tag release
+// Return cosets coefficients for circuits over BN254.
+fn coset_k() -> Vec<Fr> {
+    vec![
```

Comment:
Not a big deal, but this would be more readable if you had a vec of hex strings that you mapped over to get the field elems